File: flutter_sdk/lib/src/domain/models/chat_message/chat_message.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'package:chat_flutter_sdk/domain/models/product/product.dart';
import 'package:equatable/equatable.dart';

enum MessageRole {
  user('USER'),
  assistant('AGENT');

  final String role;

  const MessageRole(this.role);
}

enum MessageType {
  text('text'),
  image('image'),
  voice('voice'),
  product('product'),
  productCarousel('productCarousel'),
  promotion('promotion'),
  quickReply('quickReply'),

  unknown('unknown');

  final String type;
  const MessageType(this.type);
}

enum MessageStatus {
  delivered('DELIVERED'),
  read('READ'),
  error('ERROR'),
  sent('SENT'),
  inProgress('IN_PROGRESS');

  final String status;
  const MessageStatus(this.status);
}

// A class that represents a chat message in the chat
class ChatMessage extends Equatable {
  final int? id;
  final MessageRole role;
  final String content;
  final MessageType type;
  final MessageStatus status;
  final DateTime timestamp;

  // Attached file name to resource of the message
  // for images the image file path, and for audios the audio path
  final String? fileName;

  // Audio amplitudes preview
  final List<double>? amplitudes;
  // Audio duration in ms
  final int? duration;

  // Products linked to the chat message
  final List<Product> products;

  // Used to indicate if the expanded version of a message should be displayed
  // this should not be persisted in DB
  final bool expand;

  final List<String> quickReplies;

  const ChatMessage({
    this.id,
    required this.role,
    required this.type,
    required this.timestamp,
    this.content = '',
    this.status = MessageStatus.inProgress,
    this.fileName,
    this.amplitudes,
    this.duration,
    this.products = const [],
    this.expand = false,
    this.quickReplies = const [],
  });

  const ChatMessage.text({
    this.id,
    required this.role,
    required this.timestamp,
    this.status = MessageStatus.inProgress,
    required this.content,
    this.quickReplies = const [],
  }) : type = MessageType.text,
       amplitudes = null,
       fileName = null,
       duration = null,
       products = const [],
       expand = false;

  const ChatMessage.voice({
    this.id,
    required this.role,
    required this.timestamp,
    this.status = MessageStatus.inProgress,
    required this.fileName,
    required this.amplitudes,
    required this.duration,
    this.quickReplies = const [],
  }) : type = MessageType.voice,
       products = const [],
       content = '',
       expand = false;

  const ChatMessage.image({
    this.id,
    required this.role,
    required this.timestamp,
    required this.fileName,
    this.status = MessageStatus.inProgress,
    this.content = '',
    this.quickReplies = const [],
  }) : type = MessageType.image,
       amplitudes = null,
       duration = null,
       products = const [],
       expand = false;

  const ChatMessage.product({
    this.id,
    required this.role,
    required this.timestamp,
    this.status = MessageStatus.inProgress,
    this.products = const [],
    this.expand = false,
    this.quickReplies = const [],
  }) : type = MessageType.product,
       content = '',
       fileName = '',
       amplitudes = null,
       duration = null;

  const ChatMessage.carousel({
    this.id,
    required this.role,
    required this.timestamp,
    this.status = MessageStatus.inProgress,
    this.products = const [],
    this.expand = false,
    this.quickReplies = const [],
  }) : type = MessageType.productCarousel,
       content = '',
       fileName = '',
       amplitudes = null,
       duration = null;

  // Creates a copy of a chat message
  ChatMessage copyWith({
    int? id,
    MessageRole? role,
    String? content,
    MessageType? type,
    MessageStatus? status,
    String? fileName,
    List<double>? amplitudes,
    int? duration,
    List<Product>? products,
    bool? expand,
    List<String>? quickReplies,
    DateTime? timestamp,
  }) {
    return ChatMessage(
      id: id ?? this.id,
      role: role ?? this.role,
      content: content ?? this.content,
      type: type ?? this.type,
      status: status ?? this.status,
      fileName: fileName ?? this.fileName,
      amplitudes: amplitudes ?? this.amplitudes,
      duration: duration ?? this.duration,
      products: products ?? this.products,
      expand: expand ?? this.expand,
      quickReplies: quickReplies ?? this.quickReplies,
      timestamp: timestamp ?? this.timestamp,
    );
  }

  // Equatable props array
  @override
  List<Object?> get props => [
    id,
    role,
    content,
    type,
    status,
    fileName,
    amplitudes,
    duration,
    products,
    expand,
    quickReplies,
    timestamp,
  ];
}

File: android-sdk/sdk/src/main/java/ai/yalo/chat/sdk/domain/models/ChatMessage.kt
// Copyright (c) Yalochat, Inc. All rights reserved.
package ai.yalo.chat.sdk.domain.models

/** Who wrote a message. */
internal enum class MessageRole(val wireName: String) {
    User("USER"),
    Agent("AGENT"),
    ;

    companion object {
        fun of(wireName: String): MessageRole = entries.first { it.wireName == wireName }
    }
}

/**
 * What a message carries.
 *
 * The full vocabulary the backend can send, even though the SDK only renders
 * some of it so far. [Unknown] is what an unrecognised value reads back as, so
 * a newer backend cannot break an older app.
 */
internal enum class MessageType(val wireName: String) {
    Text("text"),
    Image("image"),
    Voice("voice"),
    Product("product"),
    ProductCarousel("productCarousel"),
    ProductConfirmation("productConfirmation"),
    Promotion("promotion"),
    Video("video"),
    Attachment("attachment"),
    ChatStatus("chat-status"),
    Unknown("unknown"),
    ;

    companion object {
        fun of(wireName: String): MessageType =
            entries.firstOrNull { it.wireName == wireName } ?: Unknown
    }
}

/** How far a message got. */
internal enum class MessageStatus(val wireName: String) {
    Delivered("DELIVERED"),
    Read("READ"),
    Error("ERROR"),
    Sent("SENT"),
    InProgress("IN_PROGRESS"),
    Clicked("CLICKED"),
    ;

    companion object {
        fun of(wireName: String): MessageStatus = entries.first { it.wireName == wireName }

        /** What [wireName] means, or [fallback] when it is one this SDK does not know. */
        fun of(wireName: String, fallback: MessageStatus): MessageStatus =
            entries.firstOrNull { status -> status.wireName == wireName } ?: fallback
    }
}

/** What tapping a button attached to a message does. */
internal enum class MessageButtonType(val wireName: String) {
    Reply("reply"),
    Postback("postback"),
    Link("link"),
    ;

    companion object {
        /** What [wireName] means, or [Reply] when it is one this SDK does not know. */
        fun of(wireName: String): MessageButtonType =
            entries.firstOrNull { type -> type.wireName == wireName } ?: Reply
    }
}

/**
 * A tappable option the channel attached to a message.
 *
 * [url] is only meaningful for a [MessageButtonType.Link], which is the one
 * kind that leaves the conversation.
 */
internal data class MessageButton(
    val text: String,
    val type: MessageButtonType = MessageButtonType.Reply,
    val url: String? = null,
)

/**
 * One message in a conversation.
 *
 * [id] is the local row and is null until the message is stored. [wiId] is the
 * backend's id and is null for a message the user has just written, which is
 * why the two cannot be the same field.
 *
 * [timestamp] is epoch milliseconds. It is passed in rather than read from the
 * clock so that whoever creates the message decides what time it happened.
 *
 * [voice] is the recording a [MessageType.Voice] message carries, and [image]
 * the picture a [MessageType.Image] message carries. Each is null for every
 * other kind.
 */
internal data class ChatMessage(
    val role: MessageRole,
    val type: MessageType,
    val timestamp: Long,
    val id: Long? = null,
    val wiId: String? = null,
    val content: String = "",
    val status: MessageStatus = MessageStatus.InProgress,
    val header: String? = null,
    val footer: String? = null,
    val buttons: List<MessageButton> = emptyList(),
    val voice: VoiceNote? = null,
    val image: ImageAttachment? = null,
)

/**
 * The buttons that answer the message rather than doing something else with it.
 *
 * These are what the chat offers as quick replies. Tapping one says its text
 * back to the channel, so the person picks an answer instead of typing it.
 */
internal val ChatMessage.quickReplies: List<MessageButton>
    get() = buttons.filter { button -> button.type == MessageButtonType.Reply }
